lib/vnet/escalate_windows.go

```diff
@@ -14,27 +14,361 @@
+	service, err := startService(ctx, cfg)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer service.Close()
+	log.InfoContext(ctx, "Started Windows service", "service", serviceName)
+	for {
+		select {
+		case <-ctx.Done():
+			log.InfoContext(ctx, "Context canceled, stopping Windows service")
+			if _, err := service.Control(svc.Stop); err != nil {
+				return trace.Wrap(err, "sending stop request to Windows service %s", serviceName)
+			}
+			return nil
+		case <-time.After(time.Second):
+			if status, err := service.Query(); err != nil {
+				return trace.Wrap(err, "querying admin service")
+			} else {
+				if status.State != svc.Running && status.State != svc.StartPending {
+					return trace.Errorf("service stopped running prematurely, status: %v", status)
+				}
+			}
+		}
+	}
+}
+
+// startService starts the Windows VNet admin service in the background, and
+// installs it first if necessary.
+func startService(ctx context.Context, cfg AdminProcessConfig) (*mgr.Service, error) {
+	// Avoid [mgr.Connect] because it requests elevated permissions.
+	scManager, err := windows.OpenSCManager(nil /*machine*/, nil /*database*/, windows.SC_MANAGER_CONNECT)
+	if err != nil {
+		return nil, trace.Wrap(err, "opening Windows service manager")
+	}
+	defer windows.CloseServiceHandle(scManager)
+	serviceNamePtr, err := syscall.UTF16PtrFromString(serviceName)
+	if err != nil {
+		return nil, trace.Wrap(err, "converting service name to UTF16")
+	}
+	serviceHandle, err := windows.OpenService(scManager, serviceNamePtr, serviceAccessFlags)
+	if err != nil {
+		log.InfoContext(ctx, "Failed to open Windows service, trying to install the service", "error", err)
+		if err := escalateAndInstallService(); err != nil {
+			return nil, trace.Wrap(err, "installing Windows service")
+		}
+		if serviceHandle, err = waitForServiceToInstall(ctx, scManager, serviceNamePtr); err != nil {
+			return nil, trace.Wrap(err, "waiting for service immediately after installation")
+		}
+	}
+	service := &mgr.Service{
+		Name:   serviceName,
+		Handle: serviceHandle,
+	}
+	if err := service.Start("vnet-service"); err != nil {
+		return nil, trace.Wrap(err, "starting Windows service %s", serviceName)
+	}
+	return service, nil
+}
+
+func escalateAndInstallService() error {
+	user, err := user.Current()
+	if err != nil {
+		return trace.Wrap(err, "getting current user")
+	}
+	if err := escalateAndRunSubcommand("vnet-install-service", "--userSID", user.Uid); err != nil {
+		return trace.Wrap(err, "installing VNet admin service")
+	}
+	return nil
+}
+
+func escalateAndRunSubcommand(args ...string) error {
+	tshPath, err := os.Executable()
+	if err != nil {
+		return trace.Wrap(err, "getting executable path")
+	}
+	verbPtr, err := syscall.UTF16PtrFromString("runas")
+	if err != nil {
+		return trace.Wrap(err, "converting string to UTF16")
+	}
+	pathPtr, err := syscall.UTF16PtrFromString(shsprintf.EscapeDefaultContext(tshPath))
```

Comment:
This `shsprintf.EscapeDefaultContext` is just to escape the argument passed to `ShellExecute`? I think it might be worthwhile to document this, my first thought was that it's meant to protect from some kind of injection attack, but `tshPath` comes from `os.Executable`.

FWIW, as long as we include `.exe` in the path we should be fine I think (if it even applies to `ShellExecute`, as this talks about `CreateProcess` https://flatt.tech/research/posts/batbadbut-you-cant-securely-execute-commands-on-windows/).